firedrake/assemble.py

```diff
@@ -1600,7 +1601,8 @@ def build(self):
                                 extruded=extruded,
                                 extruded_periodic=extruded_periodic,
                                 constant_layers=constant_layers,
-                                subset=self._needs_subset)
+                                subset=self._needs_subset,
+                                signature=signature)
```

Comment:
I think we do not need to pass signature here, as `kernel_args` will distinguish what we want to distinguish. I think the issue we originally had was that `_GlobalKernelBuilder` was cached keyed by `form.signature()`, by `form.signature()` did not correctly label restricted functions, which has been fixed.